### DISCUSSION TITLE
How to use custom images?

### DISCUSSION THREAD

**Discussion by andershagbard** [actions/actions-runner-controller#2085] [Questions] [answered]
Hi!

I am very new to Docker and Kubernetes, so this might be a noobish question.

I've tried to make a custom image with nodejs, npm and yarn installed. The image can be found in `andershagbard/arc`.
It's just a Dockerfile with:
```dockerfile
FROM summerwind/actions-runner:latest

RUN sudo apt-get update -y
RUN sudo apt install nodejs -y
RUN sudo apt install npm -y
RUN sudo rm -rf /var/lib/apt/lists/*

# Install Yarn
RUN sudo npm install --global yarn
```

When I try to add this image, Kubernetes keeps trying to build the runners, but always fail. It deletes the runner before I can see the error message.

How do I proceed from here? I want to use a custom image, and not use something like setup-node in my actions.

**Comment by abinmn** [comment]
Instead of running the apt in sudo, you could try setting the user to root and at the end of the file change the user to runner.'

```Dockerfile
FROM summerwind/actions-runner:latest

USER root

RUN apt-get update -y \
        && apt install nodejs npm -y \
        && rm -rf /var/lib/apt/lists/* \
        && npm install --global yarn

USER runner
```

**Reply by andershagbard** [reply]
Did not resolve my issue, but this Dockerfile is probably better than my current.

**Reply by ellisvalentiner** [reply]
I actually found that installing tools as `root` lead to permissions issues, so I set `USER runner` beforehand.

**Comment by toast-gear** [comment]
> When I try to add this image, Kubernetes keeps trying to build the runners, but always fail. It deletes the runner before I can see the error message.

I don't understand this sentence however you can use [stern](https://github.com/stern/stern) to preemptively tail the logs so you can see what is going on, it's a suggested tool in the troubleshooting guide https://github.com/actions-runner-controller/actions-runner-controller/blob/master/TROUBLESHOOTING.md#tools

**Reply by andershagbard** [reply]
I will try your tool if the above answer isn't working.

To clarify myself; Kubernetes tries to build the runner (I think). It tries to build it's container, then fails, then deletes the runner, and retries the process on a new runner.

**Comment by andershagbard** [comment]
Seems to work after i built the image with platform `linux/amd64`.

Now I am just getting permission denied error on `yarn`.

**Reply by andershagbard** [reply, marked as answer]
Got it running now.

Final file:

```dockerfile
FROM summerwind/actions-runner:latest

USER root

RUN curl -sL https://deb.nodesource.com/setup_18.x | bash - 
RUN apt-get install -y nodejs
RUN npm install --global yarn

USER runner
```

**Reply by toast-gear** [reply]
Cracking, thanks for psoting your final file! I thought I'd highlight `latest` tag may be going away and you should instead use the "latest" tag for the OS version you want `ubuntu-20.04` or `ubuntu-22.04`. At some point 20.04 image will stop being produced so I'd suggest building from 22.04. https://github.com/actions-runner-controller/actions-runner-controller/pkgs/container/actions-runner-controller%2Factions-runner https://hub.docker.com/r/summerwind/actions-runner/tags

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
